### PR TITLE
Add walk support

### PR DESCRIPTION
**Pull request by adrianheine**
This is on top of #86.

### CODE REVIEW

**Comment by RReverser** [comment]
Hmm, are many consumers really using built-in walk instead of estraverse or similar?

**Comment by adrianheine** [comment]
I'm planning on using it in bublé, but I can ship it myself. In general, I don't know what the future for acorn walk is, marijn and I had a quick talk about it in https://github.com/acornjs/acorn/issues/656#issuecomment-363814147.

**Comment by lahmatiy** [comment]
@RReverser It's quite confusing when you switch to `acorn-jsx`, attempt to walk through AST following to readme but get the error:
```
TypeError: base[type] is not a function
```
It was surprising to me that `acorn-jsx` doesn't extend the walker when it can, and I need to use "estraverse or similar" with no mention about it in `acorn` or `acorn-jsx` readmes.

I believe, since a walker is a part of `acorn` it should be extended by `acorn-jsx` as well.

File: walk.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+module.exports = function(walk) {
+  const base = walk.base;
+  base.JSXSpreadAttribute = base.SpreadElement;
+  base.JSXOpeningFragment = base.JSXEmptyExpression = base.JSXIdentifier = base.JSXText = base.JSXClosingFragment = base.Identifier;
+  base.JSXAttribute = function(node, st, c) {
+    c(node.name, st);
+    c(node.value, st);
```
Comment (lahmatiy) [inline walk.js]: `node.value` can be null – `if (node.value)` is needed here

File: walk.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  };
+  base.JSXClosingElement = function(node, st, c) {
+    c(node.name, st);
+  };
+  base.JSXElement = function(node, st, c) {
+    c(node.openingElement, st);
+    for (var i = 0; i < node.children.length; ++i) {
+      c(node.children[i], st);
+    }
+    c(node.closingElement, st);
```
Comment (lahmatiy) [inline walk.js]: `node.closingElement` can be null – `if (node.closingElement)` is needed here

File: walk.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  };
+  base.JSXFragment = function(node, st, c) {
+    c(node.openingFragment, st);
+    for (var i = 0; i < node.children.length; ++i) {
+      c(node.children[i], st);
+    }
+    c(node.closingFragment, st);
```
Comment (lahmatiy) [inline walk.js]: `node.closingFragment` can be null – `if (node.closingFragment)` is needed here

**Comment by beheh** [comment]
Looking forward to seeing this landed and released! I'd like to contribute using this in the [i18next-scanner](https://github.com/i18next/i18next-scanner) project instead of the unmaintained [acorn-jsx-walk](https://github.com/sderosiaux/acorn-jsx-walk#readme). So far all the tests in i18next-scanner pass after switching to this project/fork and adding the conditionals as described by lahmatiy.

@adrianheine do you think you could get the minor changes in this PR, or @lahmatiy just manually add them and do a release? Thanks!

**Comment by adrianheine** [comment]
Thanks for reminding me, I actually forgot this was my PR. I addressed the feedback.

File: walk.js (hunk 3 above)
Comment (adrianheine) [reply walk.js]: Can you give me an example for this case?

Comment (adrianheine) [reply walk.js]: On a syntax level, that's just `JSXElement`s. A fragment in terms of syntax is something like `<></>`.

**Comment by beheh** [comment]
Would love to see this landed and released soon! @RReverser

File: walk.js (hunk 3 above)
Comment (lahmatiy) [reply walk.js]: I see, yep, probably checking is not required here, since anonymous fragments can't have attributes.

**Comment by chacestew** [comment]
Hopeful for this to land soon. It would be a big fix for i18next-scanner.

**Comment by adrianheine** [comment]
Closing this in favor of sderosiaux/acorn-jsx-walk#3.

**Pull request closed without merging**